src/compiler/checker.ts

```diff
@@ -13798,7 +13801,12 @@ namespace ts {
 
             function inferFromSignature(source: Signature, target: Signature, skipParameters: boolean) {
                 if (!skipParameters) {
+                    const saveBivariant = bivariant;
+                    const kind = target.declaration ? target.declaration.kind : SyntaxKind.Unknown;
+                    // Once we descend into a bivariant signature we remain bivariant for all nested inferences
+                    bivariant = bivariant || kind === SyntaxKind.MethodDeclaration || kind === SyntaxKind.MethodSignature || kind === SyntaxKind.Constructor;

```

Comment:
Doesn't this only need to be set on the specifically contravariant comparison within `forEachMatchingParameterType`? Also, could we not use a single `variance` variable for both `bivariant` and `contravariant` context tracking (seeing as we have a variance enum already)?

Reply:
The `contravariant` flag continues to track whether the position is co- or contra-variant, regardless of whether we've descended into a `bivariant` position. Doing it this way, using two separate boolean variables, makes the logic that flips the flag easy. I looked at keeping it in a single variable but didn't like the complexity that came with it.